src/frontend/Canonicalize.ml

```diff
@@ -146,7 +146,7 @@ let rec no_parens {expr; emeta} =
   | Variable _ | IntNumeral _ | RealNumeral _ | ImagNumeral _ | GetLP
    |GetTarget ->
       {expr; emeta}
-  | TernaryIf _ | BinOp _ | PrefixOp _ | PostfixOp _ ->
+  | TernaryIf _ | BinOp _ | PrefixOp _ | PostfixOp _ | Promotion _ ->
       {expr= map_expression keep_parens ident expr; emeta}
```

Comment:
I don't quite understand why this is needed. Promotion is invisible so why should it keep parens? I think you need to add Promotion to the match in `keep_parens` but this `no_parens` here is puzzling.

Reply:
I made this change in response to a test output failing. If `Promotion` is in the `no_parens` branch then

```stan
parameters {
  real<lower=-1, upper=1> x_raw;
}
transformed parameters {
  real<lower=-1, upper=1> x;
  x = ((x_raw > 0) ? 1 : -1) - x_raw;
}
```
gets turned into

```stan
parameters {
  real<lower=-1, upper=1> x_raw;
}
transformed parameters {
  real<lower=-1, upper=1> x;
  x = (x_raw > 0) ? 1 : -1 - x_raw;
}
```

Which changes the meaning of the program and causes the pretty-printer to fail. I'm still not entirely sure why this is, but keeping parenthesis seemed safest in response. If you're able to play around with this model I'm sure I am missing something

Reply:
Aha, my first attempt failed because the promotion is added outside the parens. The match in `keep_parens` needs
```ocaml
and keep_parens {expr; emeta} =
  match expr with
  | Promotion (e, ut, ad) -> {expr= Promotion (keep_parens e, ut, ad); emeta}
```
Although the next line looks a bit suspicious since it removes _two_ parens at once. Shouldn't it be
```ocaml
  | Paren ({expr= Paren _; _} as e) -> keep_parens e
```

Reply:
That change to promotions does indeed fix the issue. Thanks!

I agree about the paren line - it seems like it is doing too much. Changing it to your suggestion doesn't touch any test output, which is similarly suspicious.